OpenSim/Common/DelimFileAdapter.h

```diff
@@ -400,39 +399,60 @@ DelimFileAdapter<T>::extendRead(const std::string& fileName) const {
                      _timeColumnLabel,
                      column_labels[0]);
     column_labels.erase(column_labels.begin());
-    // Set the column labels as metadata.
-    ValueArray<std::string> value_array{};
-    for(const auto& cl : column_labels)
-        value_array.upd().push_back(SimTK::Value<std::string>{cl});
-    typename TimeSeriesTable_<T>::DependentsMetaData dep_metadata{};
-    dep_metadata.setValueArrayForKey("labels", value_array);
-    table->setDependentsMetaData(dep_metadata);
 
     // Read the rows one at a time and fill up the time column container and
-    // the data container.
+    // the data container. Start with a reasonable initial capacity for
+    // tradeoff between a small file and larger files. 100 worked well for
+    // a 50 MB file with ~80000 lines.
+    std::vector<double> timeVec;
+    int initCapacity = 100;
+    int ncol = static_cast<int>(column_labels.size());
+    timeVec.reserve(initCapacity);
+    SimTK::Matrix_<T> matrix(initCapacity, ncol);
+    
+    // Initialize current row and capacity
+    int curCapacity = initCapacity;
+    int curRow = 0;
+
+    // Start looping through each line
     auto row = nextLine();
-    while(!row.empty()) {
+    while (!row.empty()) {
         ++line_num;
+        ++curRow;
+        
+        // Double capacity if we reach the end of the containers
+        if (curRow > curCapacity) {
+            curCapacity *= 2;
+            timeVec.reserve(curCapacity);
+            matrix.resizeKeep(curCapacity, ncol);
+        }
 
         // Time is column 0.
-        double time = std::stod(row.front());
+        timeVec.push_back(std::stod(row.front()));
         row.erase(row.begin());
 
         auto row_vector = readElems(row);
 
         OPENSIM_THROW_IF(row_vector.size() != column_labels.size(),
-                         RowLengthMismatch,
-                         fileName,
-                         line_num,
-                         column_labels.size(),
-                         static_cast<size_t>(row_vector.size()));
-
-        // Column 1 is time.
-        table->appendRow(time, std::move(row_vector));
+            RowLengthMismatch,
+            fileName,
+            line_num,
+            column_labels.size(),
+            static_cast<size_t>(row_vector.size()));
+        
+        matrix.updRow(curRow-1) = std::move(row_vector);
 
         row = nextLine();
     }
 
+    // Resize the matrix down to the correct number of rows
+    matrix.resizeKeep(curRow, ncol);
```

Comment:
Perhaps fine for this purpose, but it may be safer to hold onto the memory that has been allocated. If the strategy of L423 is used elsewhere (i.e., double capacity when container is full), then repeatedly adding one (e.g., in a loop) would result in the following operations for a container of size *n*: double container size to *2n*, add the row, reduce container size to *n+1*, double container size, etc. I think [Issue 401 in the simbody repo](https://github.com/simbody/simbody/issues/401) is relevant. If this is indeed what would happen, perhaps add a comment in the doxygen telling users to manually preallocate (e.g., set elements to `NaN`) rather than repeatedly appending rows.

Reply:
> it may be safer to hold onto the memory that has been allocated

I agree, but I think because of the limitations of the `Matrix_` class as reported in issue 401, there is no separation between the `size` and `capacity` of the container. Perhaps I could add a comment to change this this logic if issue 401 is fixed (possibly by just implementing this logic in simbody)?

> repeatedly adding one (e.g., in a loop) would result in the following operations for a container of size n: double container size to 2n, add the row, reduce container size to n+1, double container size, etc.

I think this indeed would happen, however, this function is limited in scope to reading in a file. So, hopefully, a user doesn't 1) add a line, 2) then write out, 3) then read in the file, in a loop? Once it's loaded in, there's still the problem that adding any rows to a TimeSeriesTable will be slow (issue 401), but I don't think that's specific to this function. Definitely let me know if I'm missing something in the interpretation here.

Reply:
> I was asking about whether the allocation strategy used here is used elsewhere

Ah, I see now. AFAIK, it's not used elsewhere.

Reply:
Ok, sounds good. Thanks- the new comments in f83a9f6 are very helpful. :+1: